src/preamble.js

```diff
@@ -419,21 +419,33 @@ function stringToAscii(str, outPtr) {
   return writeAsciiToMemory(str, outPtr, false);
 }
 
-// Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the given array that contains uint8 values, returns
+// Given a pointer 'ptr' to a UTF8-encoded string in the given array that contains uint8 values, returns
 // a copy of that string as a Javascript String object.
+// If len is passed, ptr points to an optionally null-terminated string of up to 'len' bytes.  If len is
+// not passed, ptr must point to a null-terminated string.
 
 #if TEXTDECODER
 var UTF8Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf8') : undefined;
 #endif
-function UTF8ArrayToString(u8Array, idx) {
+function UTF8ArrayToString(u8Array, idx, maxLen) {
 #if TEXTDECODER
   var endPtr = idx;
   // TextDecoder needs to know the byte length in advance, it doesn't stop on null terminator by itself.
   // Also, use the length info to avoid running tiny strings through TextDecoder, since .subarray() allocates garbage.
-  while (u8Array[endPtr]) ++endPtr;
+  // We also limit the length of the string to a maximum of 'maxLen' if it is passed.
+  while (u8Array[endPtr]) {
+    if (maxLen != undefined && (idx + maxLen) == endPtr) break;
+    ++endPtr;
+  }
```

Comment:
This can be a very performance sensitive spot, so let's instead make this read

```javascript
if (maxLen != undefined) {
  var maxPtr = idx + maxLen;
  while (endPtr < maxPtr && u8Array[endPtr]) ++endPtr;
} else {
  while(u8Array[endPtr]) ++endPtr;
}
```

so that the heavy inner loop that has to scan through the whole string length stays as slim as possible?